# Assignment: Data Wrangling

### Reading material: `tidy_data.pdf`

**Q1.** This question provides some practice cleaning variables which have common problems.
1. For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing? 

**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks (Hint: `GSAF5.xls`).

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.
2. Drop any columns that do not contain data.
3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?
4. Clean the Age variable and make a histogram of the ages of the victims.
5. What proportion of victims are male?
6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?

---

## *Phase 1: Solution without AI

Complete all questions in the following cells without generative AI. Then commit and push the
notebook before beginning Phase 2.

**Declaration:** I completed this version without generative AI.

**Student(s):** Rukaya Selim 

**Date:** Sep. 6th, 2026

In [8]:
%pip install pandas



[notice] A new release of pip is available: 26.0.1 -> 26.2.1
[notice] To update, run: python -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [11]:
import pandas as pd

airbnb = pd.read_csv("./data/airbnb_hw.csv")
airbnb.head()

print(airbnb['Price'].head(10))

airbnb['Price_clean'] = airbnb['Price'].astype(str)
airbnb['Price_clean'] = airbnb['Price_clean'].str.replace('$', '')
airbnb['Price_clean'] = airbnb['Price_clean'].str.replace(',', '')

airbnb['Price_clean'] = pd.to_numeric(airbnb['Price_clean'], errors='coerce')

missing_count = 0
for val in airbnb['Price_clean'].isna():
    if val == True:
        missing_count += 1

print("Missing prices count:")
print(missing_count)

0    145
1     37
2     28
3    199
4    549
5    149
6    250
7     90
8    270
9    290
Name: Price, dtype: str
Missing prices count:
0


In [ ]:
police = pd.read_csv("./data/mn_police_use_of_force.csv")

police['subject_injury_clean'] = police['subject_injury']

for idx in police.index:
    val = str(police.loc[idx, 'subject_injury'])
    if val in ['Yes', 'Y', 'yes']:
        police.loc[idx, 'subject_injury_clean'] = 'Yes'
    elif val in ['No', 'N', 'no']:
        police.loc[idx, 'subject_injury_clean'] = 'No'

missing_count = police['subject_injury'].isna().sum()
total_rows = len(police)
missing_prop = missing_count / total_rows

print("Proportion of missing values:", missing_prop)
police['injury_with_na'] = police['subject_injury_clean'].fillna('Missing')

crosstab_result = pd.crosstab(
    police['force_type'], 
    police['injury_with_na'], 
    margins=True
)

print("\nCross-tabulation:")
print(crosstab_result)

#If officers are less likely to record injury status under specific conditions, treating missing values as purely random introduces bias.
#When looking at the cross-tabulation, missing values occur far more frequently in lower-severity force categories compared to higher-severity incidents

Proportion of missing values: 0.7619342359767892

Cross-tabulation:
injury_with_na               Missing    No   Yes    All
force_type                                             
Baton                              2     0     2      4
Bodily Force                    7051  1093  1286   9430
Chemical Irritant               1421   131    41   1593
Firearm                            0     2     0      2
Gun Point Display                 27    33    44    104
Improvised Weapon                 74    34    40    148
Less Lethal                       87     0     0     87
Less Lethal Projectile             0     1     2      3
Maximal Restraint Technique      170     0     0    170
Police K9 Bite                    31     2    44     77
Taser                            985   150   172   1307
All                             9848  1446  1631  12925


In [ ]:
police = pd.read_csv("./data/mn_police_use_of_force.csv")

police['subject_injury_clean'] = police['subject_injury']

for idx in police.index:
    val = str(police.loc[idx, 'subject_injury'])
    if val in ['Yes', 'Y', 'yes']:
        police.loc[idx, 'subject_injury_clean'] = 'Yes'
    elif val in ['No', 'N', 'no']:
        police.loc[idx, 'subject_injury_clean'] = 'No'

missing_count = police['subject_injury'].isna().sum()
total_rows = len(police)
missing_prop = missing_count / total_rows

print("Proportion of missing values:", missing_prop)
police['injury_with_na'] = police['subject_injury_clean'].fillna('Missing')

crosstab_result = pd.crosstab(
    police['force_type'], 
    police['injury_with_na'], 
    margins=True
)

print("\nCross-tabulation:")
print(crosstab_result)

#if it is not random data that is missing, this could create bias.
#When looking at the cross-tabulation, missing values occur far more frequently in lower-severity force categories

Proportion of missing values: 0.7619342359767892

Cross-tabulation:
injury_with_na               Missing    No   Yes    All
force_type                                             
Baton                              2     0     2      4
Bodily Force                    7051  1093  1286   9430
Chemical Irritant               1421   131    41   1593
Firearm                            0     2     0      2
Gun Point Display                 27    33    44    104
Improvised Weapon                 74    34    40    148
Less Lethal                       87     0     0     87
Less Lethal Projectile             0     1     2      3
Maximal Restraint Technique      170     0     0    170
Police K9 Bite                    31     2    44     77
Taser                            985   150   172   1307
All                             9848  1446  1631  12925


## *Phase 2: AI-assisted Revision
Keep Phase 1 frozen.

- **Phase 1 commit SHA ID:** [To be provided]
- **AI tool and model used:** [To be provided]

### *Prompts used

1. [Paste prompt]
2. [Paste follow-up prompt]

### *AI-proposed changes


[Paste the AI's concise numbered list verbatim.]

1. ...
2. ...

### *Evaluation of AI suggestions


| Change | Decision | Reason and verification |
|---|---|---|
| 1 | Accept / modify / reject | ... |
| 2 | Accept / modify / reject | ... |

### *AI-assisted solution in full
After the evaluation of AI suggestions, incorporate the accepted revisions into your Phase 1 solution and provide the final Phase 2 solution in full in the following cells.

In [ ]:
# Your Phase 2 solution to the problem goes here.


### *Reflection on AI assistance
In your own words without generative AI.

[In 150–300 words, describe the main improvements, AI mistakes or limitations, how the final solution was verified, and any remaining concerns.]